ualbertalib/avalon
# Omniauth shibboleth for CCID-based authenication

Fixes #370 

Adds omniauth shibboleth for CCID-based authentication mirroring Avalon 5

Changes proposed in this pull request:

- [x] add omniauth-shibboleth gem
- [x] update `user.rb to work with shibboleth results - [link](https://github.com/ualbertalib/avalon/blob/feature/ccid-auth-370/app/models/user.rb#L90-L111) - inspired from a21578b, b6cd19d & 30d83fc
- [x] update `devise.rb` initializer
- [x] update login link in header to use shibboleth (see [OmniAuth Documentation for `user_shibboleth_omniauth_authorize_path`](https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview), [Avalon 5](https://github.com/ualbertalib/avalon/blob/master/app/views/devise/sessions/new.html.erb#L24-L29) 
 & e86bf82. 
- [x] update changelog.md - add setup requirements regarding shibboleth-sp and notes Sept 26th comment by Weiwei
- [x] retain `users/auth/identity` route


For reference, previous work in Avalon 5:

￼Initial commit: 0e1068c
Login and branding: e86bf82 (not needed - for the most part?)
Shib and email: 30d83fc
Email as User.username: b6cd19d
More shib: a21578b


To test as of 2018-11-27:

- ${SERVER_NAME}/users/auth/shibboleth
  - redirects to: UofA CCID login then one gets returned to the ERA+AV splashpage)
- ${SERVER_NAME}/Shibboleth.sso/Session
  - returns: session details see below (notice missing e-mail)

```
Miscellaneous
Session Expiration (barring inactivity): 375 minute(s)
Client Address: 129.xx.xx.xx
SSO Protocol: urn:oasis:names:tc:SAML:2.0:protocol
Identity Provider: https://${_____IST____}/saml2/idp/metadata.php
Authentication Time: yyyy-mm-ddT16:08:39Z
Authentication Context Class: urn:oasis:names:tc:SAML:2.0:ac:classes:Password
Authentication Context Decl: (none)

Attributes
eduPersonTargetedID: 1 value(s)
givenName: 1 value(s)
sn: 1 value(s)
uid: 1 value(s)
```

Setup [extracted from comment below]:
In the past, we have followed the documentation here, for the Service Provider installation, using RPM on the staging/production environments (and dev environments back then).
https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPLinuxInstall

Once it's installed. Shibd service needs to be started. And we can generate metadata from shibboleth to send to the IdP and be able to receive responses from them.




References:
- <https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview>
- <https://www.devmynd.com/blog/configuring-rails-app-single-sign-saml-multiple-providers/>
- <https://wiki.shibboleth.net/confluence/display/SHIB2/NativeSPGettingStarted>
- <https://login.ualberta.ca/saml2/idp/metadata.php>
